seurat/seuratFunctions.R

```diff
@@ -404,3 +404,158 @@ createExampleData <- function(nRow = 100, nCol = 10){
+    ix <- iy <- 0   # intersecting point
+    lineMag <- lineMagnitude(x1, y1, x2, y2)
+    if(any(lineMag < 0.00000001)) { # modified for vectorization by BAH
+      #warning("short segment")
+      #return(9999)
+      warning("At least one line segment given by x1, y1, x2, y2 is very short.")
+    }
+    u <- (((px - x1) * (x2 - x1)) + ((py - y1) * (y2 - y1)))
+    u <- u / (lineMag * lineMag)
+    if(any((u < 0.00001) || (u > 1))) { # BAH added any to vectorize
+      ## closest point does not fall within the line segment, take the shorter distance
+      ## to an endpoint
+      ix <- lineMagnitude(px, py, x1, y1)
+      iy <- lineMagnitude(px, py, x2, y2)
+      if(ix > iy)  ans <- iy
+      else ans <- ix
+    } else {
+      ## Intersecting point is on the line, use the formula
+      ix <- x1 + u * (x2 - x1)
+      iy <- y1 + u * (y2 - y1)
+      ans <- lineMagnitude(px, py, ix, iy)
+    }
+    ans
+  }
+
+  # End of helper functions by PB
+
+  ### Now for the actual findElbow function!
+
+  # Find the elbow using the method described in
+  # stackoverflow.com/a/2022348/633251
+  # but translated to R (see above).
+
+
+  y <- sort(y, decreasing = T)
+
+  # Add an index to argument values for easy plotting
+  DF <- data.frame(x = 1:length(y), y = y)
+  fit <- lm(y ~ x, DF[c(1,nrow(DF)),]) # 2 point 'fit'
+  m <- coef(fit)[2]
+  b <- coef(fit)[1]
+
+  # Check to make sure the data is concave as described
+  # in the documentation, as arbitrary trends could give
+  # misleading answers.  The following approach simply
+  # checks to make sure all values are either above or
+  # below the reference line.  This allows the values
+  # to vary quite a bit and still return an answer.
+
+  concave <- FALSE
+  use <- 2:(nrow(DF)-1)
+  refpts <- m*DF$x[use] + b
+  if (all(refpts > DF$y[use]) | all(refpts < DF$y[use])) concave <- TRUE
+  if(ignore.concavity) concave <- TRUE
+  if (!concave) stop("Your curve doesn't appear to be concave")
+
+  # Calculate the orthogonal distances
+
+  if(is.na(min.x)){
+    if(!is.na(min.y)){
+      if(!length(which(DF$y<=min.y))<1){
+        min.x = min(DF[which(DF$y<=min.y), ]$x)
+      } else {
+        print("min.y greater than smallest y")
+        min.x = 2
+      }
+    } else {
+      print("min.x and min.y are NA")
+      min.x = 2
+    }
+
+  }
+
+  use     <- min.x:(nrow(DF)-1)
+  elbowd  <- distancePointLine(DF$x[use], DF$y[use], coef(fit)[2], coef(fit)[1])
+  DF$dist <- rep(NA, nrow(DF))
+  DF$dist[use]  <- elbowd # c(NA, elbowd, NA) # first & last points don't have a distance
+
+  if (plot) {
+    edm <- which.max(DF$dist)
+    plot(DF[,1:2], type = "b", xlab = "index", ylab = "y values",
+         main = "Looking for the Elbow")
+    segments(DF$x[1], DF$y[1],
+             DF$x[nrow(DF)], DF$y[nrow(DF)], col = "red")
+    points(DF$x[edm], DF$y[edm], cex = 1.5, col = "red")
+    points(DF$x[edm], DF$y[edm], pch = 20)
+  }
+
+  if (returnIndex) return(which.max(DF$dist)) else return(DF)
```

Comment:
I'm still curious: what is the use of returning the DF if we're unable to find the elbow?  Unless I'm missing something, returning two completely different types doesnt make sense to me.  Are you expecting the caller to do some kind of "if (is.data.frame(ret)) { ...code if failed..} else {..code if success...}?

Do you think in reality we might want to call this with a "defaultElbow" argument, such that if we can infer the right # of components, use that, otherwise default to 10 or something?  If we cant infer, do you think the script should fail?